### DISCUSSION TITLE
How do I pretty-format `Value`s in a plugin?

### DISCUSSION THREAD

**Discussion by idanarye** [nushell/nushell#13080] [Q&A]
`Value` has methods for converting to a string: [`to_abbreviated_string`](https://docs.rs/nu-protocol/0.94.2/nu_protocol/enum.Value.html#method.to_abbreviated_string), [`to_debug_string`](https://docs.rs/nu-protocol/0.94.2/nu_protocol/enum.Value.html#method.to_debug_string), [`to_expanded_string`](https://docs.rs/nu-protocol/0.94.2/nu_protocol/enum.Value.html#method.to_expanded_string), and [`to_parsable_string`](https://docs.rs/nu-protocol/0.94.2/nu_protocol/enum.Value.html#method.to_parsable_string). But none of them yields the pretty tables you see in the shell itself.

Is there a way to get these formatted tables (preferably with ANSI colors) into a `String` when running a plugin?

**Comment by ghost** [comment]
the unix philosophy (and probably idea behind plugins) is that each program does one thing and allows other things to work with the result.
in nu everything returns nu data and if no one ends up using it the final result gets prettified and printed.

to add color data to columns you can copy the code from the periodic table plugin ([relevant code](https://github.com/JosephTLyons/nu_plugin_periodic_table/blob/master/src/periodic_table.rs#L24)).  
this data will still be stored as nu data with the result that its still usable by other things and will look colorful when printed.

to recreate the table output in string format you'd probably have to call the exact same code as nu does ([link](https://github.com/nushell/nushell/blob/main/crates/nu-command/src/viewers/table.rs#L123)), which will add a lot of dependencies and therefore binary size to your plugin.

**Comment by fdncred** [comment]
Typically, if you want to return a table from a plugin you'll return a list of records. That's what I do here in one of my plugins https://github.com/fdncred/nu_plugin_regex/blob/da716736d7e5ea3c26c42f07ad8fbc1fe769f0f1/src/regex_.rs#L83C5-L83C38

**Comment by idanarye** [comment]
This is my usecase: https://github.com/idanarye/nu_plugin_skim

I'm trying to run [skim](https://github.com/lotabout/skim) as a Nu plugin, so that it could filter Nu types instead of just string. This also means that the `--preview` flag, for example, would accept Nu closures:

```nushell
ls | sk --format {get name} --preview {if $in.type == "file" { bat --color always $in.name } else { ls $in.name }}
```

This works well for simple strings and for external commands where I can just capture their text:

![image](https://github.com/nushell/nushell/assets/1149255/2d8a86b2-8594-4488-aeef-cbc21c5f4d59)

But I can't handle Nu values very well. Currently I just format them with `to_expanded_string`, which looks like crap:

![image](https://github.com/nushell/nushell/assets/1149255/2e0bdb44-1aa8-42eb-a380-2b5be19482f1)

I can get semi-presentable results if I filter them as `table`s:

![image](https://github.com/nushell/nushell/assets/1149255/535fe240-e3f9-4672-8abe-c8a5186cb7af)

But:

1. Something is off - the table is not rendered properly.
2. I don't want to force the users to always pipe their preview closure through `| table`, and I've seen no way to run that command myself (I suspect that Nu's whole "no eval" thing means that plugins can't call other commands unless they are explicitly passed to them as closures)

**Comment by idanarye** [comment]
Ended up using [the `nu_table` crate](https://crates.io/crates/nu-table):

https://github.com/idanarye/nu_plugin_skim/blob/8ad909d17b21608ca662405e1145d6de7ec70229/src/nu_item.rs#L27-L46

It works - though I couldn't figure out how to get it to have colors...
